[SPARK-47558][SS] State TTL support for ValueState

sahnib opened the pull request:
**Note**: This change has been co-authored by @ericm-db  and @sahnib 

**Authors: @ericm-db @sahnib**

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR adds support for expiring state based on TTL for ValueState. Using this functionality, Spark users can specify a TTL Mode for transformWithState operator, and provide a ttlDuration/expirationTImeInMs for each value in ValueState. TTL support for List/Map State will be added in future PRs. Once the ttlDuration has expired, the value will not be returned as part of `get()` and would be cleaned up at the end of the micro-batch. 

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

These changes are needed to support TTL for ValueState. The PR supports specifying ttl for processing time or event time. Processing time ttl is calculated by adding ttlDuration to `batchTimestamp`, and event time ttl is specified using absolute expiration time (`expirationTimeInMs`). 

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, modifies the ValueState interface for specifying `ttlDuration`, and adds `ttlMode` to `transformWithState` API. 

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Added unit test cases for both event time and processing time in `ValueStateWithTTLSuite`. 

```
WARNING: Using incubator modules: jdk.incubator.foreign, jdk.incubator.vector
[info] TransformWithStateTTLSuite:
11:56:54.590 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
11:56:56.054 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate state is evicted at ttl expiry - processing time ttl (6 seconds, 244 milliseconds)
11:57:01.188 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate ttl update updates the expiration timestamp - processing time ttl (4 seconds, 465 milliseconds)
11:57:05.641 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate ttl removal keeps value in state - processing time ttl (4 seconds, 407 milliseconds)
11:57:10.041 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate multiple value states - with and without ttl - processing time ttl (3 seconds, 131 milliseconds)
11:57:13.175 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate state is evicted at ttl expiry - event time ttl (4 seconds, 186 milliseconds)
11:57:17.355 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate ttl update updates the expiration timestamp - event time ttl (4 seconds, 28 milliseconds)
11:57:21.391 WARN org.apache.spark.sql.execution.streaming.ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
[info] - validate ttl removal keeps value in state - event time ttl (4 seconds, 428 milliseconds)
11:57:25.838 WARN org.apache.spark.sql.streaming.TransformWithStateTTLSuite:

[info] Run completed in 32 seconds, 433 milliseconds.
[info] Total number of tests run: 7
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 7, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.

```

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

sahnib committed:
Add support for ValueState TTL.

sahnib committed:
Fix existing testcases, and add TTL testcase.

sahnib committed:
Improved documentation, added NERF error classes for user errors.

sahnib committed:
Regenerate docs for SQL Error conditions.

sahnib committed:
Add more testcases for value state ttl.

sahnib committed:
Fix indentation.

sahnib committed:
Add a placeholder TODO to use range scan once available.

sahnib committed:
Remove unnecessary log in TransformWithStateExec.

sahnib committed:
Rebase with latest master

sahnib committed:
Suppress method checkstyle for TTLMode.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -103,21 +110,26 @@ class StatefulProcessorHandleImpl(
 
   private var currState: StatefulProcessorHandleState = CREATED
 
-  private def verify(condition: => Boolean, msg: String): Unit = {
```
Thx for removing this

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -77,14 +78,20 @@ class StatefulProcessorHandleImpl(
+
   private val BATCH_QUERY_ID = "00000000-0000-0000-0000-000000000000"
-  private def buildQueryInfo(): QueryInfo = {
+  logInfo(s"Created StatefulProcessorHandle")
```
nit: intentional ?

sahnib replied:
Removed.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:
```diff
@@ -26,10 +26,10 @@ import org.apache.spark.sql.catalyst.expressions.{Ascending, Attribute, Expressi
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.plans.physical.Distribution
 import org.apache.spark.sql.execution._
+import org.apache.spark.sql.execution.streaming.StateKeyValueRowSchema.{KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA}
```
wonder whether we should clarify this a bit ? `StateKeyValueRowSchema` currently is only used with TWS right - should we rename the file/class to reflect that ?

sahnib replied:
Yeah, makes sense. Renamed this to `TransformWithStateKeyValueRowSchema`.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:
```diff
@@ -90,16 +92,13 @@ case class TransformWithStateExec(
       Nil
   }
 
+
```
nit: extra newline ?

sahnib replied:
Removed.

sahnib committed:
Modify Spark Connect tws API to include ttlMode.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -23,11 +23,19 @@ import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.Serializer
-  val VALUE_ROW_SCHEMA: StructType = new StructType().add("value", BinaryType)
+  val VALUE_ROW_SCHEMA: StructType = new StructType()
+    .add("value", BinaryType)
+    .add("ttlExpirationMs", LongType)
```
Would we add a long even if ttl is disabled ?

sahnib replied:
Removed logging for ttlExpiration if ttl is disabled. In order to support this functionality, I have split ValueState into ValueStateImpl and ValueStateImplWithTTL - to prevent un-necessary ttlMode checks for each function.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -62,33 +70,79 @@ class StateTypesEncoder[GK, V](
   private val rowToObjDeserializer = valExpressionEnc.resolveAndBind().createDeserializer()
   private val reusedValRow = new UnsafeRow(valEncoder.schema.fields.length)
 
+  private val NO_TTL_ENCODED_VALUE: Long = -1L
```
Could we avoid recording this explicitly ?

sahnib replied:
Done.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -77,14 +78,20 @@ class StatefulProcessorHandleImpl(
   extends StatefulProcessorHandle with Logging {
   import StatefulProcessorHandleState._
 
+  private val ttlStates: util.List[TTLState] = new util.ArrayList[TTLState]()
```
Could we add a comment for what this list is storing ?

sahnib replied:
Done

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateVariableWithTTLSupport.scala:
```diff
@@ -0,0 +1,187 @@
+/**
+ * Manages the ttl information for user state keyed with a single key (grouping key).
+ */
+class SingleKeyTTLState(
```
nit: should we call this `SingleKeyTTLStateImpl` to be consistent ?

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateVariableWithTTLSupport.scala:
```diff
@@ -0,0 +1,187 @@
+trait TTLState {
+
+  /**
+   * Perform the user state clean yp based on ttl values stored in
```
nit: `Perform the user state clean up based`

sahnib replied:
Done

sahnib committed:
Rename ttl changes scala file to TTLState.scala

sahnib committed:
Do not write ttlExpiration in valueRow if ttlMode is set to NoTTL().

sahnib committed:
Incorporated comments.

sahnib committed:
Incorporate TTLMode for initialState changes.

sahnib committed:
Modify eventTime ttl to use absolute time instead of a ttlDuration.

sahnib committed:
Use range scan encoder for ttl.

sahnib committed:
Add a testcase for multiple values (expired and unexpired) in a partition.

sahnib committed:
Add negative value check for expirationTime/ttlDuration.

sahnib committed:
Regenrate SQL error docs.

anishshri-db commented on sql/api/src/main/scala/org/apache/spark/sql/streaming/ValueState.scala:
```diff
@@ -42,8 +43,22 @@ private[sql] trait ValueState[S] extends Serializable {
+   * Update the value of the state.
+   * @param newState the new value
+   * @param ttlDuration set the ttl to current batch processing time
+   *                    (for processing time TTL mode) plus ttlDuration
```
Should we note whether we throw an exception for event time based TTL mode here ? i.e. explicitly note that only processing time TTL mode is supported for this API ?

sahnib replied:
Done.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -49,12 +54,17 @@ object StateKeyValueRowSchema {
   private val keyProjection = UnsafeProjection.create(KEY_ROW_SCHEMA)
-  private val valueProjection = UnsafeProjection.create(VALUE_ROW_SCHEMA)
+  private val valueProjection = if (hasTtl) {
+      UnsafeProjection.create(VALUE_ROW_SCHEMA_WITH_TTL)
```
nit: indent seems off ?

sahnib replied:
I indented it with 2 spaces as it was easier to read that way. I did not find any specific guidance for class variable declarations with `if .. else` in https://github.com/databricks/scala-style-guide?tab=readme-ov-file#indent.

anishshri-db replied:
Hmm - maybe its just a GH UI thing - but it appears like 4 spaces to me ?

sahnib replied:
I think its because `UnsafeProjection.create` is inside the if part. The curly brace is next line is indented with 2 spaces.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -49,12 +54,17 @@ object StateKeyValueRowSchema {
-    stateName: String) {
-  import org.apache.spark.sql.execution.streaming.StateKeyValueRowSchema._
+    stateName: String,
+    hasTtl: Boolean) extends Logging {
```
nit: do we use `Logging` somewhere ?

sahnib replied:
removed, thanks.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -17,17 +17,22 @@
 
 package org.apache.spark.sql.execution.streaming
 
+import org.apache.spark.internal.Logging
```
same here ?

sahnib replied:
removed, thanks.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -65,22 +75,49 @@ class StateTypesEncoder[GK, V](
+   * @return encoded UnsafeRow
+   */
+  def encodeSerializedGroupingKey(
+      groupingKeyBytes: Array[Byte]): UnsafeRow = {
```
nit: could we move to same line above ?

sahnib replied:
yes, done.

anishshri-db commented on sql/api/src/main/java/org/apache/spark/sql/streaming/TTLMode.java:
```diff
@@ -0,0 +1,49 @@
+@Evolving
+public class TTLMode {
+
+    /**
```
nit: not sure if indent is off here ?

sahnib replied:
Hmm, should be 2 spaces. Fixed.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala:
```diff
@@ -16,39 +16,46 @@
     stateName: String,
     keyExprEnc: ExpressionEncoder[Any],
-    valEncoder: Encoder[S]) extends ValueState[S] with Logging {
+    valEncoder: Encoder[S])
```
We can retain the older formatting ?

sahnib replied:
done

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,244 @@
+    batchTimestampMs: Option[Long],
+    eventTimeWatermarkMs: Option[Long])
+  extends ValueState[S]
+    with Logging
```
same here ? move on line above ?

sahnib replied:
Hmm. I was under the impression that multiple with needs to be intended in separate line. Moved it to one line.

anishshri-db commented:
Should we update the PR description for this - ```event time ttl is calculated by adding ttlDuration to watermarkForEviction.``` ?

anishshri-db commented on common/utils/src/main/resources/error/error-classes.json:
```diff
@@ -3565,6 +3571,19 @@
+    ],
+    "sqlState" : "42802"
+  },
+  "STATEFUL_PROCESSOR_TTL_CANNOT_BE_NEGATIVE" : {
```
nit: `TTL_VALUE_CANNOT_BE` ?

sahnib replied:
Done

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -103,22 +113,29 @@ class StatefulProcessorHandleImpl(
+      val valueStateWithTTL = new ValueStateImplWithTTL[T](store, stateName,
+        keyEncoder, valEncoder, ttlMode, batchTimestampMs, eventTimeWatermarkMs)
+
+      val ttlState = valueStateWithTTL.ttlState
```
Not sure what we want to do here ? could we just add the `valueStateWithTTL` instance to the list of `ttlStates` ?

sahnib replied:
Maybe I am not following your question. We want to add the ttlState instead of the actual state variable state to `ttlStates`.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -77,14 +78,23 @@ class StatefulProcessorHandleImpl(
+   * Stores all the active ttl states, and is used to cleanup expired values
+   * in [[doTtlCleanup()]] function.
+   */
+  private val ttlStates: util.List[TTLState] = new util.ArrayList[TTLState]()
```
is there any use-case for accessing a TTL state by name ? i.e. just wondering whether a map is better here ?

anishshri-db replied:
Probably not - but does this need to be thread-safe ?

sahnib replied:
I dont think there is a usecase to access ttl state by name. We can modify it in future if needed. 

I dont think this needs to be thread safe, as all access happens only in Task thread.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,193 @@
+    batchTimestampMs: Option[Long],
+    eventTimeWatermarkMs: Option[Long])
+  extends TTLState
+  with Logging {
```
Lets move to line aboe ?

sahnib replied:
Logging is not needed here, removed.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,193 @@
+    val iterator = store.iterator(ttlColumnFamilyName)
+    var reachedPastExpirationTime = false
+
+    while (iterator.hasNext && !reachedPastExpirationTime) {
```
Could we remove `while` and model as a `Iterator` - something similar to https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TimerStateImpl.scala#L206 ?

sahnib replied:
Maybe I am not following the suggestion. This method does not return an iterator, it instead goes through the items in ttl column family and deletes the expired values from ttl state.

anishshri-db replied:
Right - wondering if there is a more functional way to write this.AFAIK - we tend to not use `while` loops unless absolutely required. But in this case - I guess you want to terminate on condition - so could we do something like

`takeWhile(...).foreach(...)` ?

anishshri-db commented on sql/api/src/main/scala/org/apache/spark/sql/streaming/ValueState.scala:
```diff
@@ -42,8 +43,22 @@ private[sql] trait ValueState[S] extends Serializable {
+   * Update the value of the state.
+   *
+   * @param newState    the new value
+   * @param expirationMs set the ttl to expirationMs (processingTime or eventTime)
```
nit: should we rename to `expirationTimeInMs` ?

sahnib replied:
done.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -185,6 +202,16 @@ class StatefulProcessorHandleImpl(
+   * which is expired will be cleaned up from StateStore.
+   */
+  def doTtlCleanup(): Unit = {
+    ttlStates.forEach { s =>
```
Just curious - this would be used to perform eviction for all types in the future - correct ?

sahnib replied:
correct. All types should have a associated ttlState.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,193 @@
+  import org.apache.spark.sql.execution.streaming.StateTTLSchema._
+
+  private val ttlColumnFamilyName = s"_ttl_$stateName"
+  private val ttlKeyEncoder = UnsafeProjection.create(KEY_ROW_SCHEMA)
```
Could we prefix with the object name for the key and value schema ? it seems that we use the same name in a bunch of places - which might be confusing ?

sahnib replied:
Done.

sahnib committed:
Incorporate PR comments.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -77,14 +78,23 @@ class StatefulProcessorHandleImpl(
+   * Stores all the active ttl states, and is used to cleanup expired values
+   * in [[doTtlCleanup()]] function.
+   */
+  private val ttlStates: util.List[TTLState] = new util.ArrayList[TTLState]()
```
Is there any reason to store this as Java list vs Scala seq (or other equivalent data structures) ?

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImpl.scala:
```diff
@@ -16,39 +16,45 @@
   private val stateTypesEncoder = StateTypesEncoder(keySerializer, valEncoder, stateName)
+  private[sql] var ttlState: Option[SingleKeyTTLStateImpl] = None
+
+  initialize()
```
Is this change just part of refactoring ? or is it required for some reason ?

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  /** Function to check if state exists. Returns true if present and false otherwise */
+  override def exists(): Boolean = {
```
Do we need to override this ?

sahnib replied:
`ValueStateImplWithTTL` and ` ValueStateImpl` do not have a common base class, so we need to implement all functions from ValueState trait. The only common functionality was clear and exists, so I kept them separate.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,242 @@
+      override def next(): Long = {
+        val result = nextValue.get
+        nextValue = None
+
```
nit: we could remove extra line ?

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatefulProcessorHandleSuite.scala:
```diff
@@ -48,7 +49,7 @@ class StatefulProcessorHandleSuite extends StateVariableSuiteBase {
         val store = provider.getStore(0)
         val handle = new StatefulProcessorHandleImpl(store,
-          UUID.randomUUID(), keyExprEncoder, getTimeoutMode(timeoutMode))
+          UUID.randomUUID(), keyExprEncoder, TTLMode.NoTTL(), getTimeoutMode(timeoutMode))
```
Is it possible to add unit testing to `StatefulProcessorHandleSuite` or `ValueSuite` to cover the TTL functionality at a UT level ?

sahnib replied:
I added testcases for `StatefulProcessorHandleSuite`. Will add few more testcases for ValueStateImplWithTTL, and update the PR.

sahnib replied:
Added UT testcases for `ValueStateImplWithTTL` in `ValueStateSuite`.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,242 @@
+
+    if (ttlDuration != null && ttlDuration.isNegative) {
+      throw StateStoreErrors.ttlCannotBeNegative(
+        "update", stateName)
```
Done

sahnib committed:
Add additional UT for StatefulProcessorHandleSuite and incorporate PR comments.

anishshri-db commented on sql/api/src/main/java/org/apache/spark/sql/streaming/TTLMode.java:
```diff
@@ -0,0 +1,49 @@
+   * be cleaned up by Spark automatically.
+   */
+  public static final TTLMode NoTTL() {
+        return NoTTL$.MODULE$;
```
Indent seems off here again ?

anishshri-db replied:
maybe follow same format as lines below ?

sahnib replied:
Fixed.

sahnib committed:
Add UT for ValueStateImplWithTTL in ValueStateSuite.

anishshri-db commented on sql/api/src/main/scala/org/apache/spark/sql/streaming/ValueState.scala:
```diff
@@ -42,8 +43,26 @@ private[sql] trait ValueState[S] extends Serializable {
+   */
+  def update(newState: S, ttlDuration: Duration = Duration.ZERO): Unit
+
+
```
nit: extra newline ?

sahnib replied:
Removed.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,184 @@
+    ttlMode: TTLMode,
+    stateName: String,
+    store: StateStore,
+    batchTimestampMs: Option[Long],
```
Should we just pass a single value here ?

sahnib replied:
updated to pass the ttlExpirationMs.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,237 @@
+    store.createColFamilyIfAbsent(stateName, KEY_ROW_SCHEMA, VALUE_ROW_SCHEMA_WITH_TTL,
+      NoPrefixKeyStateEncoderSpec(KEY_ROW_SCHEMA))
+
+    ttlState = new SingleKeyTTLStateImpl(ttlMode, stateName, store,
```
Could we pass `this` here ?

sahnib replied:
Modified to use one trait `TTLState` and extend it from state variable.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,184 @@
+    } else if (ttlMode == TTLMode.EventTimeTTL()) {
+      eventTimeWatermarkMs.get + ttlDuration.toMillis
+    } else {
+      throw new IllegalStateException(s"cannot calculate expiration time for" +
```
Could we throw a NERF error if required ?

sahnib replied:
This code path is no longer needed.

sahnib committed:
Merge TTLState and TTLStateVariableSupport to a single trait.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:
```diff
@@ -103,22 +113,35 @@ class StatefulProcessorHandleImpl(
+    } else if (ttlMode == TTLMode.EventTimeTTL()) {
+      eventTimeWatermarkMs.get
+    } else {
+    -1
```
nit: indent seems iff again ?

Btw - should we just throw an exception here ?

sahnib replied:
We do have ttlMode validation in TransformWithStateExec, so -1 value should never be used. 

We can throw an exeception, but then we need to make this a function I think. Right now, ttlExpiraionMs is created once and used across all state variables. We cannot throw the exception while initializing the class variable.

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:
```diff
@@ -303,6 +311,244 @@ class ValueStateSuite extends StateVariableSuiteBase {
+    }
+  }
+
+  test("test Value state TTL for event time") {
```
Could we combine with test above ?

sahnib replied:
Done

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala:
```diff
@@ -0,0 +1,579 @@
+  }
+}
+
+class TransformWithStateTTLSuite
```
Could we add a suite level comment for this ?

sahnib replied:
Done.

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala:
```diff
@@ -0,0 +1,579 @@
+      expiredTimerInfo: ExpiredTimerInfo): Iterator[OutputEvent] = {
+    var results = List[OutputEvent]()
+
+    for (row <- inputRows) {
```
Could we do `foreach` here too ?

sahnib replied:
Sure

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala:
```diff
@@ -0,0 +1,579 @@
+        AdvanceManualClock(1 * 1000),
+        CheckNewAnswer(OutputEvent(ttlKey, -1, isTTLValue = true, 61000)),
+        AddData(inputStream, InputEvent(ttlKey, "get_values_in_ttl_state", -1, null)),
+        AddData(inputStream, InputEvent(noTtlKey, "get_values_in_ttl_state", -1, null)),
```
Can we also add some tests that are stopping and starting the streaming query ?

sahnib replied:
Added Stop/StartStream for existing tests to test this scenario.

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala:
```diff
@@ -0,0 +1,579 @@
+      _valueStateWithoutTTL
+    }
+
+    for (row <- inputRows) {
```
Same here ?

sahnib replied:
Done

anishshri-db commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateTTLSuite.scala:
```diff
@@ -0,0 +1,579 @@
+        AdvanceManualClock(1 * 1000),
+        CheckNewAnswer(OutputEvent("k1", -1, isTTLValue = true, 61000)),
+        AddData(inputStream, InputEvent("k1", "get_values_in_ttl_state", -1, null)),
+        AdvanceManualClock(1 * 1000),
```
Is it possible to simulate a no data batch that might still trigger state eviction for expired state ?

sahnib replied:
Added.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:
```diff
@@ -0,0 +1,150 @@
+    store.put(encodedTtlKey, EMPTY_ROW, ttlColumnFamilyName)
+  }
+
+  override def clearExpiredState(): Unit = {
```
Maybe add a small comment here ?

sahnib replied:
Done.

sahnib committed:
Incorporate PR comments.

sahnib committed:
Add TTLMode to newly introduced tests in TransformWithStateInitialStateSuite

sahnib committed:
Add TTL handling for no data batches.

sahnib committed:
Resolve conflicts with Java API, and add ttl parameter to Java testcases for TWS.

anishshri-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:
```diff
@@ -78,17 +80,15 @@ case class TransformWithStateExec(
-
-      case _ =>
-        false
+      logWarning(s"returning true from shouldRunAnotherBatch - $newInputWatermark")
```
nit: intentional ?

sahnib replied:
Removed.

anishshri-db reviewed (approved):
lgtm pending nit

sahnib committed:
Remove unnecessary log line.

sahnib commented:
@HeartSaVioR PTAL, thanks.

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:
```diff
@@ -78,7 +80,7 @@ class ValueStateSuite extends StateVariableSuiteBase {
       }
       checkError(
-        ex.asInstanceOf[SparkException],
+        ex1.asInstanceOf[SparkException],
```
Nice finding :)

HeartSaVioR reviewed (commented):
Reviewed code changes. Reviewing tests.

In overall, I'd like to understand the use case where we need to set different TTL per update. My gut feeling of main use case of state TTL was that they just consider the grouping key A to never appear again after TTL has expired (so they set an enough TTL value), which actually doesn't need such flexible TTL setup.

sahnib commented:
> Reviewed code changes. Reviewing tests.
> 
> In overall, I'd like to understand the use case where we need to set different TTL per update. My gut feeling of main use case of state TTL was that they just consider the grouping key A to never appear again after TTL has expired (so they set an enough TTL value), which actually doesn't need such flexible TTL setup.

Discussed offline, the reason we decided on two APIs because ttlDuration does not make sense for event time ttlMode. In eventTime, user might want to decide TTL based on event time column value (of row being processed), or add value to watermark (however watermark for first batch is always zero and then jumps significantly as we process first batch). Having such a interface however complicates the API. 
Its hard to decipher at this stage if Spark users would ever want to use eventTime ttl. If its needed, we should understand exact use-cases (how should ttl be calculated in event time) and then support this mode. 

Based on this discussion, we have decided to remove EventTimeTTL for now. Furthermore, to simplify the API - we accept a ttlConfig per state variable which sets ttlDuration at variable level.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -925,15 +926,15 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           keyEncoder, outputObjAttr, planLater(child), hasInitialState,
           initialStateGroupingAttrs, initialStateDataAttrs,
-          initialStateDeserializer, planLater(initialState)) :: Nil
+          initialStateDeserializer, planLater (initialState)) :: Nil
```
Removed.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -65,22 +74,48 @@ class StateTypesEncoder[GK, V](
+   * Encode the specified value in Spark UnsafeRow with no ttl.
+   * The ttl expiration will be set to -1, specifying no TTL.
+   */
   def encodeValue(value: V): UnsafeRow = {
```
The comment is outdated. I have removed it from here. Also, we do not need a default value for `expirationMs: Long` anymore. Its the whole reason I split it into 2 methods. Not sure why Scala compiler did not catch this. I have removed the default value from second parameter.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,228 @@
+    doUpdate(newState, expirationTimeInMs)
+  }
+
+  private def doUpdate(newState: S,
```
removed this method as ttlConfig is now on state variable level.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,228 @@
+
+  private def isExpired(valueRow: UnsafeRow): Boolean = {
+    val expirationMs = stateTypesEncoder.decodeTtlExpirationMs(valueRow)
+    val isExpired = expirationMs.map(
```
Actually, we could even do `expirationMs.exists(StateTTL.isExpired(_, batchTimestampMs))`. Incorporated the change,

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ValueStateImplWithTTL.scala:
```diff
@@ -0,0 +1,228 @@
+  /*
+   * Internal methods to probe state for testing. The below methods exist for unit tests
+   * to read the state ttl values, and ensure that values are persisted correctly in
+   * the underlying  state store.
```
Fixed.

sahnib commented on sql/core/src/test/java/test/org/apache/spark/sql/TestStatefulProcessor.java:
```diff
@@ -85,7 +89,7 @@ public scala.collection.Iterator<String> handleInputRows(
 
       count += numRows;
-      countState.update(count);
+      countState.update(count, Duration.ZERO);
```
As discussed offline, removed ttlDuration per update, and added ttlConfig for state variable.

sahnib commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala:
```diff
@@ -89,14 +124,29 @@ class StateTypesEncoder[GK, V](
+   * return None.
+   */
+  def decodeTtlExpirationMs(row: UnsafeRow): Option[Long] = {
+    val expirationMs = row.getLong(1)
```
I have added an assert for `hasTTL`. However, I want to note that at the end of microbatch (before `store.commit`), we will call this method multiple times for each candidate row which has expired ttl in the secondary index.

sahnib committed:
Removed EventTimeTTL, and added ttlConfig per state variable.

In eventTime TTLMode, user might want to decide TTL based on event time column value (of row being processed), or add value to watermark (however watermark for first batch is always zero and then jumps significantly as we process first batch). Having such a interface complicates the API.
Its hard to decipher at this stage if Spark users would ever want to use eventTime ttl. If its needed, we should understand exact use-cases (how should ttl be calculated in event time) and then support this mode.

Based on this discussion, we have decided to remove EventTimeTTL for now. Furthermore, to simplify the API - we accept a ttlConfig per state variable which sets ttlDuration at variable level.

sahnib committed:
Update Javadoc for TTLConfig and getValueState with ttlConfig.

HeartSaVioR reviewed (commented):
Reviewed updated code. Still reviewing test suites.

sahnib committed:
Fix ValueStateSuite and StatefulProcessorHandleSuite testcases.

sahnib commented on sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessorHandle.scala:
```diff
@@ -41,6 +41,12 @@ private[sql] trait StatefulProcessorHandle extends Serializable {
+  def getValueState[T](
+      stateName: String,
+      valEncoder: Encoder[T],
+      tTLConfig: TTLConfig): ValueState[T]
```
Fixed.

sahnib commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatefulProcessorHandleSuite.scala:
```diff
@@ -227,19 +218,17 @@ class StatefulProcessorHandleSuite extends StateVariableSuiteBase {
+        batchTimestampMs = Some(10))
 
-        val valueState = handle.getValueState("testState", Encoders.STRING)
+      val valueState = handle.getValueState("testState", Encoders.STRING)
```
It would not be, I have captured both scenarios in the test now (created a ttl and non-ttl state).

sahnib commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:
```diff
@@ -312,187 +312,109 @@ class ValueStateSuite extends StateVariableSuiteBase {
-        Encoders.STRING).asInstanceOf[ValueStateImplWithTTL[String]]
+        Encoders.STRING, ttlConfig).asInstanceOf[ValueStateImplWithTTL[String]]
       ImplicitGroupingKeyTracker.setImplicitKey("test_key")
+      testState.update("v1")
```
it should not be. I missed updating this test, somehow local run was not initially showing this as failed. Once I cleaned everything and rebuilt, failure showed up. Fixed the testcase.

sahnib commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:
```diff
@@ -312,187 +312,109 @@ class ValueStateSuite extends StateVariableSuiteBase {
+      assert(!nextBatchTestState.exists())
+      assert(nextBatchTestState.get() === null)
+
+      nextBatchTestState.clear()
```
removed the redundant check.

sahnib commented on sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:
```diff
@@ -312,187 +312,109 @@ class ValueStateSuite extends StateVariableSuiteBase {
 
-      val testState: ValueStateImplWithTTL[String] = handle.getValueState[String]("testState",
-        Encoders.STRING).asInstanceOf[ValueStateImplWithTTL[String]]
+      val nextBatchTestState: ValueStateImplWithTTL[String] = nextBatchHandle
```
Okay, on this part - we would not allow removing ttlConfig from a state variable, or adding to a variable which did not have ttl previously. (This will be enforced as part of state metadata - cc: @anishshri-db ). 

The reason to disallow is that schema is different between ValueState/ValueStateWithTTL (ttlExpiration column wont exist in first one), and mixing these 2 schemas in StateStore for a column family complicates things significantly. If a user wants to remove ttl, they can simply create a new State Variable to store future state without enforcing ttl. 

Let me know if this makes sense @HeartSaVioR

HeartSaVioR replied:
Yup, I actually expected that behavior. It is probably good to add or remove TTL functionality against existing state, but it's just good to have and maybe it's even arguable whether it's really good to have. So OK.

HeartSaVioR reviewed (approved):
+1

HeartSaVioR commented:
Thanks! Merging to master.
